src/main/java/io/fabric8/maven/docker/wait/WaitUtil.java

```diff
@@ -36,17 +42,34 @@ public static long wait(int wait, Callable<Void> callable) throws ExecutionExcep
         return delta(now);
     }
 
-    public static long wait(int maxWait, WaitChecker ... checkers) throws WaitTimeoutException {
-        return wait(maxWait, Arrays.asList(checkers));
+    public static long wait(DockerAccess access, String containerId, int maxWait, WaitChecker ... checkers) throws WaitTimeoutException, NotRunningException {
+        return wait(access, containerId, maxWait, Arrays.asList(checkers));
     }
 
-    public static long wait(int maxWait, Iterable<WaitChecker> checkers) throws WaitTimeoutException {
+    public static long wait(DockerAccess access, String containerId, int maxWait, Iterable<WaitChecker> checkers) throws WaitTimeoutException, NotRunningException {
         long max = maxWait > 0 ? maxWait : DEFAULT_MAX_WAIT;
         long now = System.currentTimeMillis();
         try {
             do {
-                if (check(checkers)) {
-                    return delta(now);
+                try {
+                    if (!access.getContainer(containerId).isRunning()) {
+                        if (check(checkers)) {
```

Comment:
This 'double checking' is actually necessary. If the container is not running anymore ,the checkers have to be run again since a checker might have changed his state. EG: If the process prints out an expected string just before it ends. In that case, the wait should just end instead of failing with a 'unexpected end' exception. 

For the same reason, i don't think it will be feasible to implement the running check as a waiter.